src/gevent/_hub_primitives.py

```diff
@@ -133,47 +130,61 @@ def __iter__(self):
     def __next__(self):
         self._begin()
 
-        if self._count == 0:
+        if not self._objects or self._count == 0:
             # Exhausted
-            self._cleanup()
-            raise StopIteration()
+            self.__exit__(None, None, None)
+            raise StopIteration
 
-        self._count -= 1
         try:
             item = self._waiter.get()
             self._waiter.clear()
             if item is self:
                 # Timer expired, no more
-                self._cleanup()
-                raise StopIteration()
+                self.__exit__(None, None, None)
+                raise StopIteration
+
+            self._count -= 1
+            self._objects.remove(item)
+            self._unlink(item)
             return item
         except:
-            self._cleanup()
+            self.__exit__(None, None, None)
```

Comment:
Let's try not to call dunder methods directly. Leave `_cleanup` as it was, call it here, and call it from `__exit__`.